network/parallelism.go

```diff
@@ -0,0 +1,263 @@
+		}
+
+		// Only record drift if we have valid values for both. There is a small window
+		// where we get a serializer value but have 0 for timestamp.:
+		if p.timestampNetworkSeconds > 0 && p.timestampSerializerSeconds > 0 {
+			p.timestampDriftSeconds = p.timestampSerializerSeconds - p.timestampNetworkSeconds
+		}
+	})
+
+	// Register serializer for tracking incming timestamp.
+	p.serializerRelease = p.statshub.RegisterSerializer(func(ss types.SerializerStats) {
+		p.mut.Lock()
+		defer p.mut.Unlock()
+
+		if ss.NewestTimestampSeconds > p.timestampSerializerSeconds {
+			p.timestampSerializerSeconds = ss.NewestTimestampSeconds
+		}
+		if p.timestampNetworkSeconds > 0 && p.timestampSerializerSeconds > 0 {
+			p.timestampDriftSeconds = p.timestampSerializerSeconds - p.timestampNetworkSeconds
+		}
+	})
+	return p
+}
+
+func (p *parallelism) Stop() {
+	p.serializerRelease()
+	p.networkRelease()
+	p.stop <- struct{}{}
+}
+
+func (p *parallelism) Run(ctx context.Context) {
+	go func() {
+		p.run(ctx)
+	}()
+}
+
+func (p *parallelism) run(ctx context.Context) {
+	p.mut.Lock()
+	p.statshub.SendParralelismStats(types.ParralelismStats{
+		MinConnections:     p.cfg.MinConnections,
+		MaxConnections:     p.cfg.MaxConnections,
+		DesiredConnections: p.currentDesired,
+	})
+	p.mut.Unlock()
+	for {
+		var checkInterval time.Duration
+		p.mut.RLock()
+		checkInterval = p.cfg.CheckInterval
+		p.mut.RUnlock()
+		select {
+		case <-ctx.Done():
+			return
+		case <-p.stop:
+			return
+		case <-time.After(checkInterval):
+			p.desiredLoops()
+		}
+	}
+}
+
+func (p *parallelism) UpdateConfig(cfg types.ParralelismConfig) {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+	p.cfg = cfg
+	p.statshub.SendParralelismStats(types.ParralelismStats{
+		MinConnections:     p.cfg.MinConnections,
+		MaxConnections:     p.cfg.MaxConnections,
+		DesiredConnections: p.currentDesired,
+	})
+}
+
+func (p *parallelism) desiredLoops() {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+
+	// Dont bother calculating if connections are the same value.
+	if p.cfg.MinConnections == p.cfg.MaxConnections {
+		level.Debug(p.l).Log("msg", "min and max loops are same no change", "desired", p.cfg.MinConnections)
+		p.changeParallelism(p.currentDesired)
+	}
+
+	// Loop over network errors and remove them if the ttl expired.
+	keepErrors := make([]time.Time, 0, len(p.networkErrors))
+	for _, err := range p.networkErrors {
+		if time.Since(err) <= p.cfg.ResetInterval {
+			keepErrors = append(keepErrors, err)
+		}
+	}
+	p.networkErrors = keepErrors
+
+	keepSuccesses := make([]time.Time, 0, len(p.networkSuccesses))
+	for _, err := range p.networkSuccesses {
+		if time.Since(err) <= p.cfg.ResetInterval {
+			keepSuccesses = append(keepSuccesses, err)
+		}
+	}
+	p.networkSuccesses = keepSuccesses
+	errorRate := p.networkErrorRate()
+	// If we have network errors then ramp down the number of loops.
+	if p.cfg.AllowedNetworkErrorPercent != 0.0 && errorRate >= p.cfg.AllowedNetworkErrorPercent {
```

Comment:
Is it always a good idea to decrease parallelism when there are network errors....? I'm not 100% sure. Maybe errors should just prevent scaling up instead of triggerring scale down?

Reply:
I think in case of downstream Prom having issues it was good to step back and give Prom some breathing room, but in case of mistakes in Alloy config or some invalid data being scraped, we may create a kind of domino effect by scaling down 🤔

Reply:
Its very possible we will have to tune it as we gain experience with it.